File: stage1/benchmarks/reference/stdlib_sync/main.rs
use std::sync::{mpsc, Mutex, OnceLock};

fn main() {
    let counter = Mutex::new(1);
    *counter.lock().unwrap() = 2;
    println!("{}", *counter.lock().unwrap());

    let ready = OnceLock::new();
    ready.set("configured").unwrap();
    println!("{}", ready.get().is_some());

    let empty: OnceLock<i32> = OnceLock::new();
    match empty.get() {
        Some(value) => println!("{value}"),
        None => println!("empty"),
    }

    let (sender, receiver) = mpsc::sync_channel(1);
    sender.send("message").unwrap();
    match receiver.try_recv() {
        Ok(message) => println!("{message}"),
        Err(_) => println!("missing"),
    }
}

File: stage1/benchmarks/reference/stdlib_sync/main.go
package main

import (
	"fmt"
	"sync"
)

func main() {
	counter := 1
	var mutex sync.Mutex
	mutex.Lock()
	counter = 2
	mutex.Unlock()
	mutex.Lock()
	fmt.Println(counter)
	mutex.Unlock()

	var ready string
	var once sync.Once
	once.Do(func() { ready = "configured" })
	fmt.Println(ready != "")

	var empty *int
	var emptyOnce sync.Once
	_ = emptyOnce
	if empty == nil {
		fmt.Println("empty")
	}

	channel := make(chan string, 1)
	channel <- "message"
	select {
	case message := <-channel:
		fmt.Println(message)
	default:
		fmt.Println("missing")
	}
}
